Allow instantiation from class methods (#498)
facebookresearch/hydra: website/docs/patterns/objects.md

        self.user = user
        self.password = password
        self.database = database
    
    @classmethod
    def default_init_method(self, database):

>         return self("localhost", "admin", "password", database)

Reviewer: I think this can be removed now.

Reviewer: I propose that you add a new section to this page describing `hydra.utils.instantiate()`.
It's probably should come near the end, after we introduce it for the first time.

├── config.yaml
└── db
    ├── mysql.yaml
    ├── postgresql.yaml

>     └── postgresql_default.yaml

Reviewer: I think we don't need a complete example.
It's enough to say in the main doc that this supports:
 - classes
 - class methods
 - static methods
 - (and if you test) static class methods.

Author: Ok, where in the main doc is this?  I can add that to this PR.

Reviewer: Does not exist yet, see my comment. we can add a new section to objects.md for now.

`cls` field and optionally use `params` to pass parameters to the function or method.

Example file for classes, class and static methods, and functions (`models.py`)
```python
class Foo:
  def __init__(x: int, y: int) -> None:
    self.x = x
    self.y = y

  @classmethod
  def class_method(self, z: int) -> Any:
    return self(z, 10)

  @staticmethod
  def static_method(z: int) -> int:
    return z + 1

def bar(z: int) -> int:
  return z + 2
```
Example configs
```yaml
# instantiates an object Foo(10, 20)
myobject:
  cls: models.Foo
  params:
    x: 10
    y: 20
# instantiates an object Foo(5, 10)
myclassmethod:
  cls: models.Foo.class_method
  params:
    z: 5
# returns 16
mystaticmethod:
  cls: models.Foo.static_method
  params:
    z: 15
# returns 17
myfunction:
  cls: models.bar
  params:
    z: 15
```
Now to test these instantiate / call them as follows:
```python
import hydra

@hydra.main(config_path="config.yaml")
def app(cfg):
  myobject = hydra.utils.call(cfg.myobject)
  myclassmethod = hydra.utils.call(cfg.myclassmethod)
  mystaticmethod = hydra.utils.call(cfg.mystaticmethod)
  myfunction = hydra.utils.call(cfg.myfunction)
```

> Note, the old `instantiate` function is now an alias of the `call` function used above.

Reviewer: ```suggestion
Note: `hydra.utils.instantiate()` is an alias for `hydra.utils.call()`. They are in fact the same function.
```

Author: done

be used to call functions and methods.  Simply put the path to the function or method in the
`cls` field and optionally use `params` to pass parameters to the function or method.

Example file for classes, class and static methods, and functions (`models.py`)
```python
class Foo:
  def __init__(x: int, y: int) -> None:
    self.x = x
    self.y = y

  @classmethod
  def class_method(self, z: int) -> Any:
    return self(z, 10)

  @staticmethod
  def static_method(z: int) -> int:
    return z + 1

def bar(z: int) -> int:
  return z + 2
```
Example configs
```yaml
# instantiates an object Foo(10, 20)
myobject:
  cls: models.Foo
  params:
    x: 10
    y: 20
# instantiates an object Foo(5, 10)
myclassmethod:
  cls: models.Foo.class_method
  params:
    z: 5
# returns 16
mystaticmethod:
  cls: models.Foo.static_method
  params:
    z: 15
# returns 17
myfunction:
  cls: models.bar
  params:
    z: 15
```
Now to test these instantiate / call them as follows:
```python
import hydra

@hydra.main(config_path="config.yaml")
def app(cfg):
  myobject = hydra.utils.call(cfg.myobject)
  myclassmethod = hydra.utils.call(cfg.myclassmethod)
  mystaticmethod = hydra.utils.call(cfg.mystaticmethod)

>   myfunction = hydra.utils.call(cfg.myfunction)

Reviewer: ```suggestion
  foo1: Foo = hydra.utils.call(cfg.myobject)
  foo2: Foo = hydra.utils.call(cfg.myclassmethod)
  ret1: int = hydra.utils.call(cfg.mystaticmethod)
  ret2: int = hydra.utils.call(cfg.myfunction)
```

$ python my_app.py db=postgresql db.params.password=abcde
PostgreSQL connecting to localhost with user=root and password=abcde and database=tutorial
```

In addition to instantiating objects from classes, the `instantiate` method can also
be used for class and static methods.  Simply put the path to the method in the `cls`

> field and use `params` if needed and instatiate your object exactly like a normal class.

Reviewer: Please describe each of the use cases in more details.
Something along those lines (suggestion, feel free to diverge):

#### class names
An object of that class will be instantiated and returned. passing the config arguments to the `__init__` method
Example class:
```python
class Foo:
  def __init__(x: int, y: int) -> None:
    self.x = x
    self.y = y
```
Example usage:
config:
```yaml
object:
  cls: Foo
  params:
    x: 10
    y: 20
```
```python
result = hydra.utils.instantiate(cfg.object)
```

#### static_method
TODO: similar small example

#### class method
TOOD: similar small example
